rig/machine_control/scp_connection.py

```diff
@@ -79,65 +125,167 @@ def send_scp(self, buffer_size, x, y, p, cmd, arg1=0, arg2=0, arg3=0,
-        while n_tries < self.n_tries:
-            # Transit the packet
-            self.sock.send(packet.bytestring)
-            n_tries += 1
+        # Send the burst
+        self.send_scp_burst(buffer_size, 1, iter(packets))
 
-            try:
-                # Try to receive the returned acknowledgement
-                ack = self.sock.recv(receive_length)
-            except IOError:
-                # There was nothing to receive from the socket
-                continue
-
-            # Convert the possible returned packet into an SCP packet. If
-            # the sequence number matches the expected sequence number then
-            # the acknowledgement has been received.
-            scp = packets.SCPPacket.from_bytestring(ack, n_args=expected_args)
-
-            # Check that the CMD_RC isn't an error
-            if scp.cmd_rc in self.error_codes:
-                raise self.error_codes[scp.cmd_rc](
-                    "Packet with arguments: cmd={}, arg1={}, arg2={}, "
-                    "arg3={}; sent to core ({},{},{}) was bad.".format(
-                        cmd, arg1, arg2, arg3, x, y, p
+        # Return the received packet
+        assert callback.packet is not None
+        return callback.packet
+
+    def send_scp_burst(self, buffer_size, window_size,
+                       parameters_and_callbacks):
+        """Send a burst of SCP packets and call a callback for each returned
+        packet.
+
+        Parameters
+        ----------
+        buffer_size : int
+            Number of bytes held in an SCP buffer by SARK, determines how many
+            bytes will be expected in a socket.
+        window_size : int
+            Number of packets which can be awaiting replies from the SpiNNaker
+            board.
+        parameters_and_callbacks: iterable of :py:class:`.scpcall`
+            Iterable of :py:class:`.scpcall` elements.  These elements can
+            specify a callback which will be called with the returned packet.
+        """
+        # Non-blocking and then the following is a busy loop.
+        self.sock.setblocking(False)
+
+        # Calculate the receive length, this should be the smallest power of
+        # two greater than the required size
+        max_length = buffer_size + consts.SDP_HEADER_LENGTH
+        receive_length = int(2**math.ceil(math.log(max_length, 2)))
+
+        class TransmittedPacket(object):
+            """A packet which has been transmitted and still awaits a response.
+            """
+            __slots__ = ["callback", "packet", "expected_args", "n_tries",
+                         "time_sent", "extra_timeout"]
+
+            def __init__(self, callback, packet, expected_args, extra_timeout):
+                self.callback = callback
+                self.packet = packet
+                self.expected_args = expected_args
+                self.extra_timeout = extra_timeout
+                self.n_tries = 1
+                self.time_sent = time.time()
```

Comment:
Wouldn't it be best to swap this field for "the time to timeout", that way you save yourself messing with working out the extra_timeout stuff every time you check it?